Update EIP-20: Renovate (#6038)
ethereum/EIPs: EIPS/eip-20.md

---
eip: 20
title: Token Standard
description: A standard interface for tokens
author: Fabian Vogelsteller <fabian@ethereum.org>, Vitalik Buterin <vitalik.buterin@ethereum.org>
discussions-to: https://github.com/ethereum/EIPs/issues/20
status: Final
type: Standards Track
category: ERC
created: 2015-11-19
---

## Abstract

> This EIP standardizes an interface for fungible tokens within smart contracts.
> The interface has basic transfer functionality, and allows tokens to be approved so they can be spent by another on-chain third party.

Reviewer: I believe this change goes beyond the minimum required to fit current standards.

```suggestion
The following standard allows for the implementation of a standard API for tokens within smart contracts.
This standard provides basic functionality to transfer tokens, as well as allow tokens to be approved so they can be spent by another on-chain third party.
```

function transferFrom(address _from, address _to, uint256 _value) public returns (bool success)
```

#### approve

Allows `_spender` to withdraw from your account multiple times, up to the `_value` amount. If this function is called again it overwrites the current allowance with `_value`.

> **NOTE**: Please read the [Security Considerations](#security-considerations) section for potential attack vectors.

Reviewer: Seems like you're removing part of the specification?

```suggestion
clients SHOULD make sure to create user interfaces in such a way that they set the allowance first to `0` before setting it to another value for the same spender.
THOUGH The contract itself shouldn't enforce it, to allow backwards compatibility with contracts deployed before

**NOTE**: Please read the [Security Considerations](#security-considerations) section for potential attack vectors.
```

Author: I didn't think so. I'll readd it if it means this can be merged.

event Approval(address indexed _owner, address indexed _spender, uint256 _value)
```

## Rationale

### The approve and transferFrom flow

`approve` and `transferFrom` can be used in conjunction with one another to allow smart contracts to take an action if and only if a transfer succeeds. This enables use cases such as decentralized exchanges, which can trustlessly swap one token for another.

### Decimals is optional

> Some tokens might wish to use a base unit other than $10^{-18}$, and sometimes it doesn't make sense to have a fraction of a token. However, not every token might reasonably need to specify `decimals`. Thus, `decimals` is optional.

Reviewer: I don't think we should invent content that wasn't there in the original proposal.

Author: It's a non-normative required section. If I were the person writing ERC-20 and decided to include this field, that would be why. If the original authors consent/agree with the motivation provided then I see nothing wrong with this.

Reviewer: This is still more than is required to comply with current guidelines, and I don't think this qualifies as errata.

## Backwards Compatibility

Many existing tokens deployed on the Ethereum network already support this EIP.

## Security Considerations

### Re-approval

#### Description of re-approval attack

1. Alice approves Bob to transfer $N$ of Alice's tokens (where $N>0$) by calling `approve(Bob, N)`
2. Later, Alice decides to set the approval from $N$ to $M$ ( $M>0$ ), so she calls `approve(Bob, M)`
3. Bob notices Alice's second transaction when it is submitted to the mempool
4. Before Alice's `approve(Bob, M)` transaction is included in a block, Bob sends a transaction calling `transferFrom(Alice, Bob, N)` with a higer priority fee than Alice's transaction, and a transaction calling `transferFrom(Alice, Bob, M)` with a lower priority fee than Alice's transaction
5. Bob's `transferFrom(Alice, Bob, N)` will be executed first, transfering $N$ tokens from Alice to Bob
6. Next, Alice's `approve(Bob, M)` will execute, allowing Bob to transfer an additonal $M$ tokens
7. Finally, Bob's `transferFrom(Alice, Bob, M)` will be executed first, transfering another $M$ tokens from Alice to Bob

Bob was able to transfer a total of $N+M$ tokens instead of a total of $M$.

> (Written by Mikhail Vladimirov <mikhail.vladimirov@gmail.com> and Dmitry Khovratovich <khovratovich@gmail.com>, edited by Pandapip1 (@Pandapip1))

Reviewer: ```suggestion
```

Generally we don't include attribution beyond the author field.
